src/modules/Game.class.js

```diff
@@ -1,68 +1,403 @@
+    idle: 'idle',
+    playing: 'playing',
+    win: 'win',
+    lose: 'lose',
+  };
+
+  constructor(
+    initialState = [
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+    ],
+  ) {
+    this.state = [...initialState];
+    this.status = Game.Status.idle;
+    this.score = 0;
+    this.initialState = initialState;
+  }
+
+  getScore() {
+    return this.score;
+  }
+
+  getState() {
+    return this.state;
+  }
+
+  getStatus() {
+    this.updateStatus();
+
+    return this.status;
+  }
+
+  start() {
+    this.status = Game.Status.playing;
+    this.state = this.initialState.concat();
+    this.createRandomTile();
+    this.createRandomTile();
+    this.printTiles();
+  }
+
+  restart() {
+    this.clearTheBoard();
+    this.printTiles();
+    this.status = Game.Status.idle;
+    this.score = 0;
+  }
+
+  moveLeft() {
+    if (this.status !== Game.Status.playing) {
+      return;
+    }
+
+    this.saveState();
+
+    let didTilesMove = false;
+
+    for (let row = 0; row < 4; row++) {
+      let didJustMerged = false;
+
+      for (let tile = 1; tile < 4; tile++) {
+        if (this.state[row][tile] > 0) {
+          let moveTile = 0;
+
+          let merge = false;
+
+          while (this.state[row][tile - moveTile - 1] === 0) {
+            moveTile++;
+          }
+
+          if (
+            !didJustMerged &&
+            this.state[row][tile - moveTile - 1] === this.state[row][tile]
+          ) {
+            moveTile++;
+            merge = true;
+          }
+
+          if (moveTile > 0) {
+            if (!didJustMerged && merge) {
+              didJustMerged = true;
+
+              const newNumber = this.state[row][tile - moveTile] * 2;
+
+              this.state[row][tile - moveTile] = newNumber;
+
+              this.score += newNumber;
+            } else {
+              didJustMerged = false;
+
+              this.state[row][tile - moveTile] = this.state[row][tile];
+            }
+
+            this.state[row][tile] = 0;
+
+            didTilesMove = true;
+          }
+        }
+      }
+    }
+
+    if (didTilesMove) {
+      this.createRandomTile();
+    }
+
+    return didTilesMove;
+  }
+
+  moveRight() {
+    if (this.status !== Game.Status.playing) {
+      return;
+    }
+
+    this.saveState();
+
+    let didTilesMove = false;
+
+    for (let row = 0; row < 4; row++) {
+      let didJustMerged = false;
+
+      for (let tile = 2; tile >= 0; tile--) {
+        if (this.state[row][tile] > 0) {
+          let moveTile = 0;
+          let merge = false;
+
+          while (this.state[row][tile + moveTile + 1] === 0) {
+            moveTile++;
+          }
+
+          if (
+            !didJustMerged &&
+            this.state[row][tile + moveTile + 1] === this.state[row][tile]
+          ) {
+            moveTile++;
+            merge = true;
+          }
+
+          if (moveTile > 0) {
+            if (!didJustMerged && merge) {
+              didJustMerged = true;
+
```

Comment:
This logic is quite similar and it is duplicated in every method, try to move it to a helper function and use it everywhere